### ISSUE TITLE
NuGet installed test adapters breaks NUnit on mono

### ISSUE THREAD

**Issue by iskiselev** [nunit/nunit-vs-adapter#77]
If in project output folder presence any of "nunit.core.dll", "nunit.core.interfaces.dll" or "nunit.util.dll" (and they will if project has installed test adapter from NuGet), than attempt to run any test from that folder under mono will fail with InvalidCastException.

**Comment by CharliePoole** [comment]
Can you give more precise instructions for reproducing this? That is, what steps do you take to run under mono?

**Comment by iskiselev** [comment]
Please look on example:
https://github.com/iskiselev/TravisExperiments/commit/3dd10829ba98beb459f13fce51a3393d969cd381
Here is sample of running this application on mono:
https://travis-ci.org/iskiselev/TravisExperiments/builds/76875707
Please look on next log lines:
```
[0K$ mono ./testrunner/NUnit.Runners.2.6.4/tools/nunit-console.exe ./bin/TravisExperiments.dll
WARNING: The runtime version supported by this application is unavailable.
Using default runtime: v4.0.30319
NUnit-Console version 2.6.4.14350
Copyright (C) 2002-2012 Charlie Poole.

Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.

Copyright (C) 2000-2002 Philip Craig.

All Rights Reserved.

Runtime Environment - 
   OS Version: Unix 3.13.0.40
  CLR Version: 4.0.30319.17020 ( Mono 4.0 ( 4.0.3 (Stable 4.0.3.20/d6946b4 Tue Aug  4 09:43:57 UTC 2015) ) )

ProcessModel: Default    DomainUsage: Single
Execution Runtime: mono-4.0
Unhandled Exception:
System.InvalidCastException: Cannot cast from source type to destination type.

Server stack trace: 
  at (wrapper xdomain-dispatch) NUnit.Core.RemoteTestRunner:Load (object,byte[]&,byte[]&)

Exception rethrown at [0]: 

  at (wrapper xdomain-invoke) NUnit.Core.RemoteTestRunner:Load (NUnit.Core.TestPackage)
  at NUnit.Util.TestDomain.Load (NUnit.Core.TestPackage package) [0x00000] in <filename unknown>:0 
travis_time:end:02979cab:start=1440351695306066495,finish=1440351695623284465,duration=317217970
[0K
[31;1mThe command "mono ./testrunner/NUnit.Runners.2.6.4/tools/nunit-console.exe ./bin/TravisExperiments.dll" exited with 156.[0m
travis_time:start:20c7823c
[0K$ rm bin/nunit.core.dll
travis_time:end:20c7823c:start=1440351695626905941,finish=1440351695630482674,duration=3576733
[0K
[32;1mThe command "rm bin/nunit.core.dll" exited with 0.[0m
travis_time:start:01d9aa98
[0K$ rm bin/nunit.core.interfaces.dll
travis_time:end:01d9aa98:start=1440351695633992716,finish=1440351695637427931,duration=3435215
[0K
[32;1mThe command "rm bin/nunit.core.interfaces.dll" exited with 0.[0m
travis_time:start:003828b3
[0K$ rm bin/nunit.util.dll
travis_time:end:003828b3:start=1440351695640932866,finish=1440351695644434570,duration=3501704
[0K
[32;1mThe command "rm bin/nunit.util.dll" exited with 0.[0m
travis_time:start:03d6d087
[0K$ mono ./testrunner/NUnit.Runners.2.6.4/tools/nunit-console.exe ./bin/TravisExperiments.dll
WARNING: The runtime version supported by this application is unavailable.
Using default runtime: v4.0.30319
NUnit-Console version 2.6.4.14350
Copyright (C) 2002-2012 Charlie Poole.

Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.

Copyright (C) 2000-2002 Philip Craig.

All Rights Reserved.

Runtime Environment - 
   OS Version: Unix 3.13.0.40
  CLR Version: 4.0.30319.17020 ( Mono 4.0 ( 4.0.3 (Stable 4.0.3.20/d6946b4 Tue Aug  4 09:43:57 UTC 2015) ) )

ProcessModel: Default    DomainUsage: Single
Execution Runtime: mono-4.0
.Test executed

Tests run: 1, Errors: 0, Failures: 0, Inconclusive: 0, Time: 0.01374 seconds
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
```

There I've run tests from sample project using nunit-console.exe and it fails with System.InvalidCastException, after it I've deleted "nunit.core.dll", "nunit.core.interfaces.dll" or "nunit.util.dll" and run it again - and everything passed sucessful.

You can reproduce same on windows if you install mono, and then run any nunit test runner with mono.

**Comment by CharliePoole** [comment]
There seems to be a combination of issues causing your problem.

NUnit 2.x uses the 2.0 runtime although it can (and does) run under higher runtimes. You are using Mono 4.0, which no longer supports the 2.0 runtime, so everything gets run under the higher level runtime. That's what's happening with you.

NUnit 2.x has always had an issue with it's internal assemblies being referenced by the tests. We actually __expect__ those assemblies to be not found and their loading is handled by an AssemblyLoad event. I hadn't heard of a problem before simply because the assemblies are present, without being referenced. Check if you have references.

The Nuget installation creates references to the assemblies, and leaves them as CopyLocal = true. We then set CopyLocal=false in a powershell script. If you did the nuget installation on a Linux system, then the powershell script wouldn't run, so that may be a contributing factor. You should change CopyLocal.

So there seem to be two issues: 1) how are these assemblies ending up in your output directory and 2) exactly why that's a problem.

**Comment by iskiselev** [comment]
Thank you for clarification, looks like setting all referenced libraries as not for copy local resolve issue.
I've checked once more - if I install package under VS2015, it does not set CopyLocal=false.
It works good in VS2013.

**Comment by iskiselev** [comment]
According to http://blog.nuget.org/20150729/Introducing-nuget-uwp.html and https://docs.nuget.org/release-notes/nuget-3.1 support of install.ps1 is deprericated and was removed in NuGet 3.1.
Maybe it worse to ask question in NuGet/Home#627 how to support developer only libraries dependencies (copylocal=false) without powershell install script.

**Comment by CharliePoole** [comment]
Thanks for asking them.

**Comment by iskiselev** [comment]
Tested it more - install.ps1 is called, but $package contains only two properties - Id and Version. There is no AssemblyReferences inside it, that's why script do nothing.

**Comment by OsirisTerje** [comment]
I had a conversation with the nuget team about this, and it is a bug in Nuget 3, see https://github.com/NuGet/Home/issues/1245

**Comment by CharliePoole** [comment]
So let's go back to the original issue and see what remains. 

How are the NUnit assemblies getting onto the Travis system? Are you checking them into your repo or is your script restoring them? Either one will be a problem when trying to build the project on a Linux system, if the assemblies end up in the output directory.

**Comment by iskiselev** [comment]
Really this issue will be resolved as soon as #78 will be resolved.
I have next scenario: project was created in Visual Studio 2015, after it NUnitTestAdapter packege was installed. Due to #78, nunit assemblies was not marked with CopyLocal=False.
After it project was saved and checked into repository. On Unix side, packages was restored (no script run at this point and it is right) and project built. On this step nunit assemblies were placed to output folder, as they were explicitly referenced by project with CopyLocal=True.

There is still small problem, that if somebody will install NUnitTestAdapter in non-Visual Studio environment (as install.ps1 will not run at all or will not have DTE), that user may have problems with running test case, but I don't think that anybody will install NUnitTestAdapter out of Visual Studio environment.

It still may worse to check why nunit fails if assemblies are in output folder, but you also could ignore it - as it is already incorrect behavior. Probably nunit test runners itself may do some more verbose output for this problem.

So, I think you could close this issue - thank you for your help.

**Comment by CharliePoole** [comment]
I'd say it's not generally a good idea to check the adapter into the repo for any application that runs on non-windows systems. If you can find an alternative, I'd do that. One way would be to use the vsix.

In any case, I see it the same way and I'll close this issue.
